{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ecd558c6",
   "metadata": {},
   "source": [
    "# Summarizing Data Post QC\n",
    "\n",
    "Author: Zan Koenig"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d5d8fb3",
   "metadata": {},
   "source": [
    "## Index\n",
    "1. [Set Default Paths](#1.-Set-Default-Paths)\n",
    "2. [Read in Pre-QC Dataset and Apply Quality Control Filters](#2.-Read-in-Pre-QC-Dataset-and-Apply-Quality-Control-Filters)\n",
    "3. [Organize the Dataset](#3.-Organize-the-Dataset)\n",
    "4. [Annotate Table with Relatedness Information](#4.-Annotate-Table-with-Relatedness-Information)\n",
    "5. [Calculate Statistics per Population](#5.-Calculate-Statistics-per-Population)\n",
    "6. [Format Table for Exporting](#6.-Format-Table-for-Exporting)\n",
    "7. [Export Final Table](#7.-Export-Final-Table)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f921f711",
   "metadata": {},
   "source": [
    "# General Overview:\n",
    "\n",
    "The purpose of this script is to format and write out a tsv which will be used to create plots and summaries of the post-QC dataset in R.\n",
    "\n",
    "**This script contains information on how to:**\n",
    "- Select specific columns from a matrix table (shortened as mt)\n",
    "- Annotate filter flags onto a matrix table\n",
    "- Join the columns of two matrix tables\n",
    "- Join two tables\n",
    "- Group a matrix table by region, population \n",
    "- Use `hl.agg.stats` to calculate statics for a metric within a population\n",
    "- Count the number of samples where a filter flag is `True`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a187324a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import hail as hl\n",
    "\n",
    "# Functions from gnomAD library to apply genotype filters   \n",
    "from gnomad.utils.filtering import filter_to_adj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5418c66a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running on Apache Spark version 3.1.3\n",
      "SparkUI available at http://mty-m.c.diverse-pop-seq-ref.internal:36703\n",
      "Welcome to\n",
      "     __  __     <>__\n",
      "    / /_/ /__  __/ /\n",
      "   / __  / _ `/ / /\n",
      "  /_/ /_/\\_,_/_/_/   version 0.2.109-b71b065e4bb6\n",
      "LOGGING: writing to /home/hail/hail-20230317-1455-0.2.109-b71b065e4bb6.log\n"
     ]
    }
   ],
   "source": [
    "# Initializing Hail \n",
    "hl.init()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8d704d35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.jp-OutputArea-output pre {\n",
       "  overflow-y: auto;\n",
       "  max-height: 500px;\n",
       "}</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Allow output scrolling in Jupyter nb viewer for cells with long outputs \n",
    "\n",
    "from IPython.core.display import HTML\n",
    "css = open('format.css').read()\n",
    "HTML('<style>{}</style>'.format(css))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd8d4f55",
   "metadata": {},
   "source": [
    "# 1. Set Default Paths\n",
    "\n",
    "These default paths can be edited by users as needed. It is recommended to run these tutorials without writing out datasets.\n",
    "\n",
    "**By default, all of the dataset write out sections are shown as markdown cells. If you would like to write out your own dataset, you can copy the code and paste it into a new code cell. Don't forget to change the paths in the following cell accordingly.** \n",
    "\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b79fb149",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path for HGDP+1kGP dataset prior to applying gnomAD QC filters\n",
    "pre_qc_path = 'gs://gcp-public-data--gnomad/release/3.1.2/mt/genomes/gnomad.genomes.v3.1.2.hgdp_1kg_subset_dense.mt'\n",
    "\n",
    "# PCA outliers file \n",
    "outliers_path = 'gs://gcp-public-data--gnomad/release/3.1/secondary_analyses/hgdp_1kg/pca/pca_outliers.txt'\n",
    "\n",
    "# Paths to related and unrelated Matrix Tables (without outliers) written out in Notebook 2: PCA and Ancestry Analyses\n",
    "unrelateds_path = 'gs://gcp-public-data--gnomad/release/3.1/secondary_analyses/hgdp_1kg/pca_results/unrelateds_without_outliers.mt'\n",
    "relateds_path = 'gs://gcp-public-data--gnomad/release/3.1/secondary_analyses/hgdp_1kg/pca_results/relateds_without_outliers.mt'\n",
    "\n",
    "# Path for final output table in tsv format\n",
    "final_table_path = 'gs://gcp-public-data--gnomad/release/3.1/secondary_analyses/hgdp_1kg/metadata_and_qc/post_qc_summary.tsv'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caebffde",
   "metadata": {},
   "source": [
    "# 2. Read in Pre-QC Dataset and Apply Quality Control Filters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a60924b8",
   "metadata": {},
   "source": [
    "Since the post-QC mt was not written out, we run the same function as tutorial notebook 1 and 2 to apply the quality control filters to the pre-QC dataset.\n",
    "\n",
    "**To avoid errors, make sure to run the next two cells before running any code that includes the post-QC dataset.**\n",
    "\n",
    "**If running the cell below results in an error, double check that you used the  `--packages gnomad` argument when starting your cluster.**  \n",
    "\n",
    "- See the tutorials [README](https://github.com/atgu/hgdp_tgp/tree/master/tutorials#readme) for more information on how to start a cluster.\n",
    "\n",
    "<br>\n",
    "<details><summary> For more information on Hail methods and expressions click <u><span style=\"color:blue\">here</span></u>.</summary> \n",
    "<ul>\n",
    "<li><a href=\"https://hail.is/docs/0.2/methods/impex.html#hail.methods.read_matrix_table\"> More on  <i> read_matrix_table() </i></a></li>\n",
    "\n",
    "<li><a href=\"https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable.count\"> More on  <i> count() </i></a></li>\n",
    "    \n",
    "<li><a href=\"https://hail.is/docs/0.2/linalg/hail.linalg.BlockMatrix.html#hail.linalg.BlockMatrix.filter_cols\"> More on  <i> filter_cols() </i></a></li>\n",
    "\n",
    "<li><a href=\"https://hail.is/docs/0.2/linalg/hail.linalg.BlockMatrix.html#hail.linalg.BlockMatrix.filter_rows\"> More on  <i> filter_rows() </i></a></li>\n",
    "</ul>\n",
    "</details>\n",
    "\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4d370069",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up function to apply gnomAD's sample, variant and genotype QC filters\n",
    "\n",
    "def run_qc(mt):\n",
    "    \n",
    "    ## Apply sample QC filters to dataset \n",
    "    # This filters to only samples that passed gnomAD's sample QC hard filters  \n",
    "    mt = mt.filter_cols(~mt.gnomad_sample_filters.hard_filtered) # removed 31 samples\n",
    "    \n",
    "    ## Apply variant QC filters to dataset\n",
    "    # This subsets to only PASS variants - those which passed gnomAD's variant QC\n",
    "    # PASS variants have an entry in the filters field \n",
    "    mt = mt.filter_rows(hl.len(mt.filters) != 0, keep=False)\n",
    "\n",
    "    ## Apply genotype QC filters to the dataset\n",
    "    # This is done using a function imported from gnomAD and is the last step in the QC process\n",
    "    mt = filter_to_adj(mt)\n",
    "\n",
    "    return mt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9dc5697f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the HGDP+1kGP pre-QC mt\n",
    "pre_qc_mt = hl.read_matrix_table(pre_qc_path)\n",
    "\n",
    "# Run QC \n",
    "mt = run_qc(pre_qc_mt)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "017ce194",
   "metadata": {},
   "source": [
    "# 3. Organize the Dataset\n",
    "\n",
    "Here we are walking through some steps to get the post-QC dataset ready for downstream analyses. We first create a table with only sample data and then select only the columns we need for the table we will write out. Next, we create a checkpoint of that table to speed up downstream analyses.\n",
    "\n",
    "<br>\n",
    "<details><summary>For more information on Hail methods and expressions click <u><span style=\"color:blue\">here</span></u>.</summary> \n",
    "<br>\n",
    "<li><a href=\"https://hail.is/docs/0.2/utils/index.html#hail.utils.hadoop_open\"> More on  <i> hadoop_open() </i></a></li>\n",
    "\n",
    "<li><a href=\"https://hail.is/docs/0.2/methods/genetics.html#hail.methods.sample_qc\"> More on  <i> sample_qc() </i></a></li>\n",
    "    \n",
    "<li><a href=\"https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable.cols\"> More on  <i> cols() </i></a></li>\n",
    "    \n",
    "<li><a href=\"https://hail.is/docs/0.2/hail.Table.html#hail.Table.select\"> More on  <i> select() </i></a></li>    \n",
    "    \n",
    "<li><a href=\"https://hail.is/docs/0.2/methods/impex.html#hail.methods.read_table\"> More on  <i> read_table() </i></a></li>\n",
    "\n",
    "<li><a href=\"https://hail.is/docs/0.2/hail.Table.html#hail.Table.count\"> More on  <i> count() </i></a></li>\n",
    "    \n",
    "<li><a href=\"https://hail.is/docs/0.2/hail.Table.html#hail.Table.checkpoint\"> More on  <i> checkpoint() </i></a></li>\n",
    "     \n",
    "</details>\n",
    "\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ff406cf8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(159795273, 4096)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove PCA outliers from the dataset\n",
    "\n",
    "# Use hl.hadoop_open to read in the PCA outliers file into Hail from Google Cloud Storage\n",
    "with hl.utils.hadoop_open(outliers_path) as file:\n",
    "    outliers = [line.rstrip('\\n') for line in file]\n",
    "\n",
    "# Use hl.literal to convert the outliers list from a python object to a Hail expression so that it can be used to filter out samples\n",
    "outliers_list = hl.literal(outliers)\n",
    "\n",
    "# Keep the samples which are not contained in the pca outlier list\n",
    "mt_without_outliers = mt.filter_cols(~outliers_list.contains(mt['s']))\n",
    "\n",
    "mt_without_outliers.count() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "de3979fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-03-03 19:55:35.224 Hail: WARN: cols(): Resulting column table is sorted by 'col_key'.\n",
      "    To preserve matrix table column order, first unkey columns with 'key_cols_by()'\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "4096"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grab the column fields of the Matrix Table \n",
    "mt_col_table = mt_without_outliers.cols()\n",
    "\n",
    "# Write a col table with only the columns needed for table 1\n",
    "mt_col_table = mt_col_table.select(mt_col_table.hgdp_tgp_meta.genetic_region,\n",
    "                             mt_col_table.hgdp_tgp_meta.population,\n",
    "                             mt_col_table.sample_qc.n_snp, \n",
    "                             mt_col_table.bam_metrics.mean_coverage)\n",
    "\n",
    "# Validity check - there should be 4096 samples \n",
    "mt_col_table.count()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c03b40c3",
   "metadata": {},
   "source": [
    "# 4. Annotate Table with Relatedness Information\n",
    "\n",
    "Relatedness information is added to the dataset so we can filter out related individuals. \n",
    "\n",
    "<br>\n",
    "<details><summary>For more information on Hail methods and expressions click <u><span style=\"color:blue\">here</span></u>.</summary> \n",
    "    \n",
    "<br>\n",
    "<li><a href=\"https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable.annotate_cols\"> More on  <i> annotate_cols() </i></a></li>\n",
    "\n",
    "<li><a href=\"https://hail.is/docs/0.2/aggregators.html#hail.expr.aggregators.counter\"> More on  <i> counter() </i></a></li>\n",
    "    \n",
    "<li><a href=\"https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable.union_cols\"> More on  <i> union_cols() </i></a></li>\n",
    "    \n",
    "</details>\n",
    "\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "c6bb8e99",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3378 718\n"
     ]
    }
   ],
   "source": [
    "# Need to get number of unrelateds annotated to the table\n",
    "\n",
    "# Read in the unrelated and related matrix tables which were written out in Notebook 2: PCA and Ancestry Analyses\n",
    "unrelateds = hl.read_matrix_table(unrelateds_path)\n",
    "relateds = hl.read_matrix_table(relateds_path)\n",
    "\n",
    "# Annotate both the unrelated and the related tables with a flag named 'unrelated' \n",
    "# Set the unrelated flag to 'True' for those in the unrelated dataset, and 'False' for those in the related dataset\n",
    "unrelateds = unrelateds.annotate_cols(unrelated=True)\n",
    "relateds = relateds.annotate_cols(unrelated=False)\n",
    "\n",
    "# Use hl.cols() to obtain two tables with only the columns from the unrelated and related mts\n",
    "unrelateds_cols = unrelateds.cols()\n",
    "relateds_cols = relateds.cols()\n",
    "\n",
    "# Validity check \n",
    "print(unrelateds_cols.count(), relateds_cols.count()) # 3378 unrelated and 718 related samples = 4096 total samples "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "afb1a29c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unrelated individuals per population: frozendict({'ACB': 94, 'ASW': 52, 'Adygei': 17, 'BEB': 99, 'Balochi': 23, 'BantuKenya': 10, 'BantuSouthAfrica': 8, 'Basque': 23, 'Bedouin': 42, 'BiakaPygmy': 22, 'Brahui': 23, 'Burusho': 24, 'CDX': 88, 'CEU': 120, 'CHB': 103, 'CHS': 103, 'CLM': 94, 'Cambodian': 9, 'Colombian': 3, 'Dai': 9, 'Daur': 9, 'Druze': 35, 'ESN': 103, 'FIN': 98, 'French': 27, 'GBR': 87, 'GIH': 100, 'GWD': 116, 'Han': 43, 'Hazara': 16, 'Hezhen': 8, 'IBS': 104, 'ITU': 102, 'Italian': 11, 'JPT': 102, 'Japanese': 29, 'KHV': 101, 'Kalash': 21, 'Karitiana': 10, 'LWK': 91, 'Lahu': 5, 'MSL': 83, 'MXL': 63, 'Makrani': 22, 'Mandenka': 20, 'Maya': 19, 'MbutiPygmy': 12, 'Melanesian': 11, 'Miao': 10, 'Mongola': 10, 'Mozabite': 25, 'Naxi': 8, 'Orcadian': 14, 'Oroqen': 8, 'PEL': 85, 'PJL': 97, 'PUR': 104, 'Palestinian': 38, 'Papuan': 17, 'Pathan': 24, 'Pima': 11, 'Russian': 25, 'STU': 98, 'San': 6, 'Sardinian': 27, 'She': 9, 'Sindhi': 22, 'Surui': 7, 'TSI': 103, 'Tu': 10, 'Tujia': 10, 'Tuscan': 8, 'Uygur': 6, 'Xibo': 9, 'YRI': 117, 'Yakut': 25, 'Yi': 10, 'Yoruba': 21})\n",
      "\n",
      "Number of related individuals per population: frozendict({'ACB': 20, 'ASW': 19, 'BEB': 32, 'BantuKenya': 2, 'Bedouin': 3, 'BiakaPygmy': 4, 'CDX': 5, 'CEU': 56, 'CHS': 60, 'CLM': 36, 'Cambodian': 1, 'Colombian': 4, 'Druze': 7, 'ESN': 45, 'GBR': 3, 'GIH': 3, 'GWD': 60, 'Hazara': 4, 'Hezhen': 1, 'IBS': 47, 'ITU': 4, 'KHV': 21, 'Kalash': 2, 'Karitiana': 2, 'LWK': 6, 'Lahu': 3, 'MSL': 15, 'MXL': 34, 'Makrani': 1, 'Mandenka': 3, 'Maya': 2, 'MbutiPygmy': 2, 'Melanesian': 2, 'Mozabite': 1, 'Naxi': 1, 'Orcadian': 1, 'PEL': 37, 'PJL': 48, 'PUR': 34, 'Palestinian': 7, 'Pima': 3, 'STU': 16, 'She': 1, 'Sindhi': 1, 'Surui': 1, 'YRI': 58})\n"
     ]
    }
   ],
   "source": [
    "# Annotate the unrelated and related mts with counts per population\n",
    "unrelateds_count = unrelateds_cols.aggregate(hl.agg.counter(unrelateds_cols.hgdp_tgp_meta.population))\n",
    "relateds_count = relateds_cols.aggregate(hl.agg.counter(relateds_cols.hgdp_tgp_meta.population))\n",
    "\n",
    "# Validity check - print out the number of unrelated and related individuals per population \n",
    "print(f\"Number of unrelated individuals per population: \\\n",
    "{unrelateds_count}\\n\\nNumber of related individuals per population: {relateds_count}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "227dc2cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "frozendict({False: 718, True: 3378})\n"
     ]
    }
   ],
   "source": [
    "# Join the columns of the unrelated and related datasets\n",
    "mt_combined = unrelateds.union_cols(relateds)\n",
    "\n",
    "# Validity check - count the number of unrelateds (True values) in the mt to make sure it is as expected\n",
    "print(mt_combined.aggregate_cols(hl.agg.counter(mt_combined.unrelated))) # 3378 True and 718 False\n",
    "\n",
    "# Create a table with only the columns from the mt containing related information\n",
    "# This is done since the final output will be a tsv and thus must be in table format\n",
    "# Being a table of columns allows it to be annotated onto the existing mt_col_table as shown below\n",
    "mt_combined_col_table = mt_combined.cols()\n",
    "\n",
    "# Annotate the relatedness information onto the column table\n",
    "mt_col_table = mt_col_table.annotate(unrelated = mt_combined_col_table[mt_col_table.s].unrelated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "32bd333b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-11-18 21:04:21 Hail: INFO: Coerced sorted dataset\n",
      "2022-11-18 21:04:21 Hail: INFO: Coerced sorted dataset\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table><thead><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td></tr><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">s</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">genetic_region</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">population</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">n_snp</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">mean_coverage</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">unrelated</div></td></tr><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">bool</td></tr>\n",
       "</thead><tbody><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;HG00096&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;EUR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;GBR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2566022</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.29e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;HG00097&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;EUR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;GBR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2569966</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.15e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;HG00099&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;EUR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;GBR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2567942</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.64e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;HG00100&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;EUR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;GBR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2576696</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.02e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;HG00101&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;EUR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;GBR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2565175</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.28e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td></tr>\n",
       "</tbody></table><p style=\"background: #fdd; padding: 0.4em;\">showing top 5 rows</p>\n"
      ],
      "text/plain": [
       "+-----------+----------------+------------+---------+---------------+\n",
       "| s         | genetic_region | population |   n_snp | mean_coverage |\n",
       "+-----------+----------------+------------+---------+---------------+\n",
       "| str       | str            | str        |   int64 |       float64 |\n",
       "+-----------+----------------+------------+---------+---------------+\n",
       "| \"HG00096\" | \"EUR\"          | \"GBR\"      | 2566022 |      3.29e+01 |\n",
       "| \"HG00097\" | \"EUR\"          | \"GBR\"      | 2569966 |      3.15e+01 |\n",
       "| \"HG00099\" | \"EUR\"          | \"GBR\"      | 2567942 |      3.64e+01 |\n",
       "| \"HG00100\" | \"EUR\"          | \"GBR\"      | 2576696 |      3.02e+01 |\n",
       "| \"HG00101\" | \"EUR\"          | \"GBR\"      | 2565175 |      3.28e+01 |\n",
       "+-----------+----------------+------------+---------+---------------+\n",
       "\n",
       "+-----------+\n",
       "| unrelated |\n",
       "+-----------+\n",
       "|      bool |\n",
       "+-----------+\n",
       "|      True |\n",
       "|      True |\n",
       "|      True |\n",
       "|      True |\n",
       "|      True |\n",
       "+-----------+\n",
       "showing top 5 rows"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mt_col_table.show(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e5a35e5",
   "metadata": {},
   "source": [
    "# 5. Calculate Statistics per Population\n",
    "\n",
    "In this section, we will be using `hl.agg.stats()` which calculates the following metrics for a given expression:\n",
    "- min\n",
    "- max\n",
    "- mean\n",
    "- standard deviation\n",
    "- number of non-missing records\n",
    "- sum\n",
    " \n",
    "Using `hl.group_by()`, we calculate these statistics for each of the 78 populations in this dataset.\n",
    "We also count the number of related samples within each populations by using `hl.agg.count_where()` and count the number of times the field denoting if samples are related or not is True.\n",
    "\n",
    "<br>\n",
    "<details><summary>For more information on Hail methods and expressions click <u><span style=\"color:blue\">here</span></u>.</summary> \n",
    "<br>\n",
    "<li><a href=\"https://hail.is/docs/0.2/aggregators.html#hail.expr.aggregators.stats\"> More on  <i> stats() </i></a></li>\n",
    "\n",
    "<li><a href=\"https://hail.is/docs/0.2/hail.Table.html#hail.Table.group_by\"> More on  <i> group_by() </i></a></li>\n",
    "    \n",
    "<li><a href=\"https://hail.is/docs/0.2/aggregators.html#hail.expr.aggregators.count_where\"> More on  <i> count_where() </i></a></li>\n",
    "    \n",
    "<li><a href=\"https://hail.is/docs/0.2/hail.expr.TupleExpression.html#hail.expr.TupleExpression.show\"> More on <i>show()</i></a></li>\n",
    "     \n",
    "</details>\n",
    "\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e6e68964",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate stats per population for each metric grouped by genetic region and population\n",
    "table = n_snp = mt_col_table.group_by(\n",
    "    mt_col_table.genetic_region, mt_col_table.population).aggregate(\n",
    "    n_snp_stats = hl.agg.stats(mt_col_table.n_snp),\n",
    "    cov_stats = hl.agg.stats(mt_col_table.mean_coverage),\n",
    "    n_unrelated = hl.agg.count_where(mt_col_table.unrelated))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "c8ef63bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-11-18 21:08:46 Hail: INFO: Coerced sorted dataset\n",
      "2022-11-18 21:08:46 Hail: INFO: Coerced sorted dataset\n",
      "2022-11-18 21:08:47 Hail: INFO: Ordering unsorted dataset with network shuffle\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table><thead><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"6\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"6\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td></tr><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"6\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">n_snp_stats</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"6\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">cov_stats</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td></tr><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">genetic_region</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">population</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">mean</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">stdev</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">min</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">max</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">n</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">sum</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">mean</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">stdev</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">min</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">max</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">n</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">sum</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">n_unrelated</div></td></tr><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int64</td></tr>\n",
       "</thead><tbody><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ACB&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.14e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.63e+04</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.92e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.20e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">114</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.58e+08</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.17e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.48e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.84e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.20e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">114</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.62e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">94</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ASW&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.08e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6.23e+04</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.88e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.18e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">71</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.18e+08</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.23e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.36e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.75e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">5.21e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">71</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.29e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">52</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;BantuKenya&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.18e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.01e+04</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.14e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.21e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">12</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.82e+07</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.19e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.88e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.91e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.40e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">12</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.83e+02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">10</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;BantuSouthAfrica&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.27e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.15e+04</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.21e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.35e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.62e+07</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.89e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">1.03e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.05e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6.45e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.11e+02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;BiakaPygmy&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.41e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">9.15e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.39e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.43e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">26</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8.86e+07</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.23e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.05e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.71e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.01e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">26</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8.41e+02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">22</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ESN&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.20e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8.46e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.18e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.23e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">148</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.74e+08</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.23e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.23e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.80e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">5.35e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">148</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.78e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">103</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;GWD&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.19e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">1.19e+04</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.15e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.22e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">176</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">5.61e+08</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.25e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.71e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.82e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.34e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">176</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">5.72e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">116</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;LWK&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.19e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">9.91e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.15e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.22e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">97</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.10e+08</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.29e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.32e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.79e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6.02e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">97</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.19e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">91</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;MSL&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.23e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8.52e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.20e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.26e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">98</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.17e+08</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.18e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.69e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.75e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.56e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">98</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.12e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">83</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;Mandenka&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.18e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">9.50e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.16e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.21e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">23</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">7.32e+07</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.24e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.65e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.73e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.02e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">23</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">7.45e+02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">20</td></tr>\n",
       "</tbody></table><p style=\"background: #fdd; padding: 0.4em;\">showing top 10 rows</p>\n"
      ],
      "text/plain": [
       "+----------------+--------------------+------------------+-------------------+\n",
       "| genetic_region | population         | n_snp_stats.mean | n_snp_stats.stdev |\n",
       "+----------------+--------------------+------------------+-------------------+\n",
       "| str            | str                |          float64 |           float64 |\n",
       "+----------------+--------------------+------------------+-------------------+\n",
       "| \"AFR\"          | \"ACB\"              |         3.14e+06 |          4.63e+04 |\n",
       "| \"AFR\"          | \"ASW\"              |         3.08e+06 |          6.23e+04 |\n",
       "| \"AFR\"          | \"BantuKenya\"       |         3.18e+06 |          2.01e+04 |\n",
       "| \"AFR\"          | \"BantuSouthAfrica\" |         3.27e+06 |          4.15e+04 |\n",
       "| \"AFR\"          | \"BiakaPygmy\"       |         3.41e+06 |          9.15e+03 |\n",
       "| \"AFR\"          | \"ESN\"              |         3.20e+06 |          8.46e+03 |\n",
       "| \"AFR\"          | \"GWD\"              |         3.19e+06 |          1.19e+04 |\n",
       "| \"AFR\"          | \"LWK\"              |         3.19e+06 |          9.91e+03 |\n",
       "| \"AFR\"          | \"MSL\"              |         3.23e+06 |          8.52e+03 |\n",
       "| \"AFR\"          | \"Mandenka\"         |         3.18e+06 |          9.50e+03 |\n",
       "+----------------+--------------------+------------------+-------------------+\n",
       "\n",
       "+-----------------+-----------------+---------------+-----------------+\n",
       "| n_snp_stats.min | n_snp_stats.max | n_snp_stats.n | n_snp_stats.sum |\n",
       "+-----------------+-----------------+---------------+-----------------+\n",
       "|         float64 |         float64 |         int64 |         float64 |\n",
       "+-----------------+-----------------+---------------+-----------------+\n",
       "|        2.92e+06 |        3.20e+06 |           114 |        3.58e+08 |\n",
       "|        2.88e+06 |        3.18e+06 |            71 |        2.18e+08 |\n",
       "|        3.14e+06 |        3.21e+06 |            12 |        3.82e+07 |\n",
       "|        3.21e+06 |        3.35e+06 |             8 |        2.62e+07 |\n",
       "|        3.39e+06 |        3.43e+06 |            26 |        8.86e+07 |\n",
       "|        3.18e+06 |        3.23e+06 |           148 |        4.74e+08 |\n",
       "|        3.15e+06 |        3.22e+06 |           176 |        5.61e+08 |\n",
       "|        3.15e+06 |        3.22e+06 |            97 |        3.10e+08 |\n",
       "|        3.20e+06 |        3.26e+06 |            98 |        3.17e+08 |\n",
       "|        3.16e+06 |        3.21e+06 |            23 |        7.32e+07 |\n",
       "+-----------------+-----------------+---------------+-----------------+\n",
       "\n",
       "+----------------+-----------------+---------------+---------------+\n",
       "| cov_stats.mean | cov_stats.stdev | cov_stats.min | cov_stats.max |\n",
       "+----------------+-----------------+---------------+---------------+\n",
       "|        float64 |         float64 |       float64 |       float64 |\n",
       "+----------------+-----------------+---------------+---------------+\n",
       "|       3.17e+01 |        2.48e+00 |      2.84e+01 |      4.20e+01 |\n",
       "|       3.23e+01 |        3.36e+00 |      2.75e+01 |      5.21e+01 |\n",
       "|       3.19e+01 |        3.88e+00 |      2.91e+01 |      4.40e+01 |\n",
       "|       3.89e+01 |        1.03e+01 |      3.05e+01 |      6.45e+01 |\n",
       "|       3.23e+01 |        3.05e+00 |      2.71e+01 |      4.01e+01 |\n",
       "|       3.23e+01 |        3.23e+00 |      2.80e+01 |      5.35e+01 |\n",
       "|       3.25e+01 |        2.71e+00 |      2.82e+01 |      4.34e+01 |\n",
       "|       3.29e+01 |        4.32e+00 |      2.79e+01 |      6.02e+01 |\n",
       "|       3.18e+01 |        2.69e+00 |      2.75e+01 |      4.56e+01 |\n",
       "|       3.24e+01 |        2.65e+00 |      2.73e+01 |      4.02e+01 |\n",
       "+----------------+-----------------+---------------+---------------+\n",
       "\n",
       "+-------------+---------------+-------------+\n",
       "| cov_stats.n | cov_stats.sum | n_unrelated |\n",
       "+-------------+---------------+-------------+\n",
       "|       int64 |       float64 |       int64 |\n",
       "+-------------+---------------+-------------+\n",
       "|         114 |      3.62e+03 |          94 |\n",
       "|          71 |      2.29e+03 |          52 |\n",
       "|          12 |      3.83e+02 |          10 |\n",
       "|           8 |      3.11e+02 |           8 |\n",
       "|          26 |      8.41e+02 |          22 |\n",
       "|         148 |      4.78e+03 |         103 |\n",
       "|         176 |      5.72e+03 |         116 |\n",
       "|          97 |      3.19e+03 |          91 |\n",
       "|          98 |      3.12e+03 |          83 |\n",
       "|          23 |      7.45e+02 |          20 |\n",
       "+-------------+---------------+-------------+\n",
       "showing top 10 rows"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Check that each of the table fields contain what we'd expect\n",
    "table.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84f67bf5",
   "metadata": {},
   "source": [
    "# 6. Format Table for Exporting\n",
    "\n",
    "In this section, we format the table before exporting so it is in a usable format once written out. Specifically, we are flattening the table. This is done so that when the table is written out, the data is easier to work with. If the tables were written out without flattening them, the new annotated information would be in a nested structure which would make it difficult to work with outside of Hail. \n",
    "\n",
    "<br>\n",
    "<details><summary>For more information on Hail methods and expressions click <u><span style=\"color:blue\">here</span></u>.</summary> \n",
    "    \n",
    "<br>\n",
    "<li><a href=\"https://hail.is/docs/0.2/hail.Table.html#hail.Table.flatten\"> More on  <i> flatten() </i></a></li>\n",
    "    \n",
    "<li><a href=\"https://hail.is/docs/0.2/hail.Table.html#hail.Table.key_by\"> More on <i>key_by()</i></a></li>\n",
    "    \n",
    "<li><a href=\"https://hail.is/docs/0.2/hail.Table.html#hail.Table.describe\"> More on <i>describe()</i></a></li>\n",
    "    \n",
    "</details>\n",
    "\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "d6706fec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flatten out the structs created from annotating the tables\n",
    "table = table.flatten()\n",
    "\n",
    "# Change the keys of the table so that it is keyed by genetic region and population\n",
    "table = table.key_by(table.genetic_region, table.population)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "d6f541b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-11-18 21:09:47 Hail: INFO: Coerced sorted dataset\n",
      "2022-11-18 21:09:47 Hail: INFO: Coerced sorted dataset\n",
      "2022-11-18 21:09:48 Hail: INFO: Ordering unsorted dataset with network shuffle\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table><thead><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td></tr><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">genetic_region</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">population</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">n_snp_stats.mean</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">n_snp_stats.stdev</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">n_snp_stats.min</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">n_snp_stats.max</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">n_snp_stats.n</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">n_snp_stats.sum</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">cov_stats.mean</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">cov_stats.stdev</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">cov_stats.min</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">cov_stats.max</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">cov_stats.n</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">cov_stats.sum</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">n_unrelated</div></td></tr><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int64</td></tr>\n",
       "</thead><tbody><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ACB&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.14e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.63e+04</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.92e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.20e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">114</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.58e+08</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.17e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.48e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.84e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.20e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">114</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.62e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">94</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ASW&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.08e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6.23e+04</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.88e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.18e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">71</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.18e+08</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.23e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.36e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.75e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">5.21e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">71</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.29e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">52</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;BantuKenya&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.18e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.01e+04</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.14e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.21e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">12</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.82e+07</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.19e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.88e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.91e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.40e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">12</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.83e+02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">10</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;BantuSouthAfrica&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.27e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.15e+04</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.21e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.35e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.62e+07</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.89e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">1.03e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.05e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6.45e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.11e+02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;BiakaPygmy&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.41e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">9.15e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.39e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.43e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">26</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8.86e+07</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.23e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.05e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.71e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.01e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">26</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8.41e+02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">22</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ESN&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.20e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8.46e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.18e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.23e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">148</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.74e+08</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.23e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.23e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.80e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">5.35e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">148</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.78e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">103</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;GWD&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.19e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">1.19e+04</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.15e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.22e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">176</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">5.61e+08</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.25e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.71e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.82e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.34e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">176</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">5.72e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">116</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;LWK&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.19e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">9.91e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.15e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.22e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">97</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.10e+08</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.29e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.32e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.79e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6.02e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">97</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.19e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">91</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;MSL&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.23e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8.52e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.20e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.26e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">98</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.17e+08</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.18e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.69e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.75e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.56e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">98</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.12e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">83</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;AFR&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;Mandenka&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.18e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">9.50e+03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.16e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.21e+06</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">23</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">7.32e+07</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3.24e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.65e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2.73e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4.02e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">23</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">7.45e+02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">20</td></tr>\n",
       "</tbody></table><p style=\"background: #fdd; padding: 0.4em;\">showing top 10 rows</p>\n"
      ],
      "text/plain": [
       "+----------------+--------------------+------------------+-------------------+\n",
       "| genetic_region | population         | n_snp_stats.mean | n_snp_stats.stdev |\n",
       "+----------------+--------------------+------------------+-------------------+\n",
       "| str            | str                |          float64 |           float64 |\n",
       "+----------------+--------------------+------------------+-------------------+\n",
       "| \"AFR\"          | \"ACB\"              |         3.14e+06 |          4.63e+04 |\n",
       "| \"AFR\"          | \"ASW\"              |         3.08e+06 |          6.23e+04 |\n",
       "| \"AFR\"          | \"BantuKenya\"       |         3.18e+06 |          2.01e+04 |\n",
       "| \"AFR\"          | \"BantuSouthAfrica\" |         3.27e+06 |          4.15e+04 |\n",
       "| \"AFR\"          | \"BiakaPygmy\"       |         3.41e+06 |          9.15e+03 |\n",
       "| \"AFR\"          | \"ESN\"              |         3.20e+06 |          8.46e+03 |\n",
       "| \"AFR\"          | \"GWD\"              |         3.19e+06 |          1.19e+04 |\n",
       "| \"AFR\"          | \"LWK\"              |         3.19e+06 |          9.91e+03 |\n",
       "| \"AFR\"          | \"MSL\"              |         3.23e+06 |          8.52e+03 |\n",
       "| \"AFR\"          | \"Mandenka\"         |         3.18e+06 |          9.50e+03 |\n",
       "+----------------+--------------------+------------------+-------------------+\n",
       "\n",
       "+-----------------+-----------------+---------------+-----------------+\n",
       "| n_snp_stats.min | n_snp_stats.max | n_snp_stats.n | n_snp_stats.sum |\n",
       "+-----------------+-----------------+---------------+-----------------+\n",
       "|         float64 |         float64 |         int64 |         float64 |\n",
       "+-----------------+-----------------+---------------+-----------------+\n",
       "|        2.92e+06 |        3.20e+06 |           114 |        3.58e+08 |\n",
       "|        2.88e+06 |        3.18e+06 |            71 |        2.18e+08 |\n",
       "|        3.14e+06 |        3.21e+06 |            12 |        3.82e+07 |\n",
       "|        3.21e+06 |        3.35e+06 |             8 |        2.62e+07 |\n",
       "|        3.39e+06 |        3.43e+06 |            26 |        8.86e+07 |\n",
       "|        3.18e+06 |        3.23e+06 |           148 |        4.74e+08 |\n",
       "|        3.15e+06 |        3.22e+06 |           176 |        5.61e+08 |\n",
       "|        3.15e+06 |        3.22e+06 |            97 |        3.10e+08 |\n",
       "|        3.20e+06 |        3.26e+06 |            98 |        3.17e+08 |\n",
       "|        3.16e+06 |        3.21e+06 |            23 |        7.32e+07 |\n",
       "+-----------------+-----------------+---------------+-----------------+\n",
       "\n",
       "+----------------+-----------------+---------------+---------------+\n",
       "| cov_stats.mean | cov_stats.stdev | cov_stats.min | cov_stats.max |\n",
       "+----------------+-----------------+---------------+---------------+\n",
       "|        float64 |         float64 |       float64 |       float64 |\n",
       "+----------------+-----------------+---------------+---------------+\n",
       "|       3.17e+01 |        2.48e+00 |      2.84e+01 |      4.20e+01 |\n",
       "|       3.23e+01 |        3.36e+00 |      2.75e+01 |      5.21e+01 |\n",
       "|       3.19e+01 |        3.88e+00 |      2.91e+01 |      4.40e+01 |\n",
       "|       3.89e+01 |        1.03e+01 |      3.05e+01 |      6.45e+01 |\n",
       "|       3.23e+01 |        3.05e+00 |      2.71e+01 |      4.01e+01 |\n",
       "|       3.23e+01 |        3.23e+00 |      2.80e+01 |      5.35e+01 |\n",
       "|       3.25e+01 |        2.71e+00 |      2.82e+01 |      4.34e+01 |\n",
       "|       3.29e+01 |        4.32e+00 |      2.79e+01 |      6.02e+01 |\n",
       "|       3.18e+01 |        2.69e+00 |      2.75e+01 |      4.56e+01 |\n",
       "|       3.24e+01 |        2.65e+00 |      2.73e+01 |      4.02e+01 |\n",
       "+----------------+-----------------+---------------+---------------+\n",
       "\n",
       "+-------------+---------------+-------------+\n",
       "| cov_stats.n | cov_stats.sum | n_unrelated |\n",
       "+-------------+---------------+-------------+\n",
       "|       int64 |       float64 |       int64 |\n",
       "+-------------+---------------+-------------+\n",
       "|         114 |      3.62e+03 |          94 |\n",
       "|          71 |      2.29e+03 |          52 |\n",
       "|          12 |      3.83e+02 |          10 |\n",
       "|           8 |      3.11e+02 |           8 |\n",
       "|          26 |      8.41e+02 |          22 |\n",
       "|         148 |      4.78e+03 |         103 |\n",
       "|         176 |      5.72e+03 |         116 |\n",
       "|          97 |      3.19e+03 |          91 |\n",
       "|          98 |      3.12e+03 |          83 |\n",
       "|          23 |      7.45e+02 |          20 |\n",
       "+-------------+---------------+-------------+\n",
       "showing top 10 rows"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Check format of the flattened table\n",
    "table.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "e2ac8e0c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------\n",
      "Global fields:\n",
      "    'global_annotation_descriptions': struct {\n",
      "        gnomad_sex_imputation_ploidy_cutoffs: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        gnomad_population_inference_pca_metrics: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        sample_hard_filter_cutoffs: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        gnomad_sample_qc_metric_outlier_cutoffs: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        gnomad_age_distribution: struct {\n",
      "            Description: str, \n",
      "            sub_globals: struct {\n",
      "                bin_edges: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                bin_freq: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_smaller: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_larger: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        hgdp_tgp_freq_meta: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        gnomad_freq_meta: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        hgdp_tgp_freq_index_dict: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        gnomad_freq_index_dict: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        gnomad_faf_meta: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        gnomad_faf_index_dict: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        variant_filtering_model: struct {\n",
      "            Description: set<str>, \n",
      "            sub_globals: struct {\n",
      "                model_name: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                score_name: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                snv_cutoff: struct {\n",
      "                    Description: str, \n",
      "                    sub_globals: struct {\n",
      "                        bin: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        min_score: struct {\n",
      "                            Description: str\n",
      "                        }\n",
      "                    }\n",
      "                }, \n",
      "                indel_cutoff: struct {\n",
      "                    Description: str, \n",
      "                    sub_globals: struct {\n",
      "                        bin: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        min_score: struct {\n",
      "                            Description: str\n",
      "                        }\n",
      "                    }\n",
      "                }, \n",
      "                snv_training_variables: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                indel_training_variables: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        variant_inbreeding_coeff_cutoff: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        vep_version: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        vep_csq_header: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        dbsnp_version: struct {\n",
      "            Description: str\n",
      "        }\n",
      "    } \n",
      "    'sample_annotation_descriptions': struct {\n",
      "        s: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        bam_metrics: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                pct_bases_20x: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                pct_chimeras: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                freemix: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                mean_coverage: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                median_coverage: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                mean_insert_size: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                median_insert_size: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                pct_bases_10x: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        sample_qc: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                n_deletion: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_het: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_hom_ref: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_hom_var: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_insertion: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_non_ref: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_snp: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_transition: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_transversion: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                r_het_hom_var: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                r_insertion_deletion: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                r_ti_tv: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        gnomad_sex_imputation: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                chr20_mean_dp: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                chrX_mean_dp: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                chrY_mean_dp: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                chrX_ploidy: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                chrY_ploidy: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                X_karyotype: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                Y_karyotype: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                sex_karyotype: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                f_stat: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_called: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                expected_homs: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                observed_homs: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        gnomad_population_inference: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                pca_scores: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                pop: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                prob_afr: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                prob_ami: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                prob_amr: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                prob_asj: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                prob_eas: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                prob_fin: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                prob_mid: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                prob_nfe: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                prob_oth: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                prob_sas: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        gnomad_sample_qc_residuals: struct {\n",
      "            Description: tuple (\n",
      "                str\n",
      "            ), \n",
      "            sub_annotations: struct {\n",
      "                n_snp_residual: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                r_ti_tv_residual: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                r_insertion_deletion_residual: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_insertion_residual: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_deletion_residual: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                r_het_hom_var_residual: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_transition_residual: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_transversion_residual: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        gnomad_sample_filters: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                hard_filters: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                hard_filtered: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                release_related: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                qc_metrics_filters: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        gnomad_high_quality: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        gnomad_release: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        relatedness_inference: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                related_samples: struct {\n",
      "                    Description: str, \n",
      "                    sub_annotations: struct {\n",
      "                        s: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        kin: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        ibd0: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        ibd1: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        ibd2: struct {\n",
      "                            Description: str\n",
      "                        }\n",
      "                    }\n",
      "                }, \n",
      "                related: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        hgdp_tgp_meta: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                project: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                study_region: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                population: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                genetic_region: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                latitude: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                longitude: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                hgdp_technical_meta: struct {\n",
      "                    Description: str, \n",
      "                    sub_annotations: struct {\n",
      "                        source: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        library_type: struct {\n",
      "                            Description: str\n",
      "                        }\n",
      "                    }\n",
      "                }, \n",
      "                global_pca_scores: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                subcontinental_pca: struct {\n",
      "                    Description: str, \n",
      "                    sub_annotations: struct {\n",
      "                        pca_scores: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        pca_scores_outliers_removed: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        outlier: struct {\n",
      "                            Description: str\n",
      "                        }\n",
      "                    }\n",
      "                }, \n",
      "                gnomad_labeled_subpop: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        high_quality: struct {\n",
      "            Description: str\n",
      "        }\n",
      "    } \n",
      "    'gnomad_sex_imputation_ploidy_cutoffs': struct {\n",
      "        x_ploidy_cutoffs: struct {\n",
      "            upper_cutoff_X: float64, \n",
      "            lower_cutoff_XX: float64, \n",
      "            upper_cutoff_XX: float64, \n",
      "            lower_cutoff_XXX: float64\n",
      "        }, \n",
      "        y_ploidy_cutoffs: struct {\n",
      "            lower_cutoff_Y: float64, \n",
      "            upper_cutoff_Y: float64, \n",
      "            lower_cutoff_YY: float64\n",
      "        }, \n",
      "        f_stat_cutoff: float64\n",
      "    } \n",
      "    'gnomad_population_inference_pca_metrics': struct {\n",
      "        n_pcs: int32, \n",
      "        min_prob: float64\n",
      "    } \n",
      "    'sample_hard_filter_cutoffs': struct {\n",
      "        min_cov: int32, \n",
      "        max_n_snp: float64, \n",
      "        min_n_snp: float64, \n",
      "        max_n_singleton: float64, \n",
      "        max_r_het_hom_var: float64, \n",
      "        max_pct_contamination: float64, \n",
      "        max_pct_chimera: float64, \n",
      "        min_median_insert_size: int32\n",
      "    } \n",
      "    'gnomad_sample_qc_metric_outlier_cutoffs': struct {\n",
      "        lms: struct {\n",
      "            n_snp: struct {\n",
      "                beta: array<float64>, \n",
      "                standard_error: array<float64>, \n",
      "                t_stat: array<float64>, \n",
      "                p_value: array<float64>, \n",
      "                multiple_standard_error: float64, \n",
      "                multiple_r_squared: float64, \n",
      "                adjusted_r_squared: float64, \n",
      "                f_stat: float64, \n",
      "                multiple_p_value: float64, \n",
      "                n: int32\n",
      "            }, \n",
      "            n_singleton: struct {\n",
      "                beta: array<float64>, \n",
      "                standard_error: array<float64>, \n",
      "                t_stat: array<float64>, \n",
      "                p_value: array<float64>, \n",
      "                multiple_standard_error: float64, \n",
      "                multiple_r_squared: float64, \n",
      "                adjusted_r_squared: float64, \n",
      "                f_stat: float64, \n",
      "                multiple_p_value: float64, \n",
      "                n: int32\n",
      "            }, \n",
      "            r_ti_tv: struct {\n",
      "                beta: array<float64>, \n",
      "                standard_error: array<float64>, \n",
      "                t_stat: array<float64>, \n",
      "                p_value: array<float64>, \n",
      "                multiple_standard_error: float64, \n",
      "                multiple_r_squared: float64, \n",
      "                adjusted_r_squared: float64, \n",
      "                f_stat: float64, \n",
      "                multiple_p_value: float64, \n",
      "                n: int32\n",
      "            }, \n",
      "            r_insertion_deletion: struct {\n",
      "                beta: array<float64>, \n",
      "                standard_error: array<float64>, \n",
      "                t_stat: array<float64>, \n",
      "                p_value: array<float64>, \n",
      "                multiple_standard_error: float64, \n",
      "                multiple_r_squared: float64, \n",
      "                adjusted_r_squared: float64, \n",
      "                f_stat: float64, \n",
      "                multiple_p_value: float64, \n",
      "                n: int32\n",
      "            }, \n",
      "            n_insertion: struct {\n",
      "                beta: array<float64>, \n",
      "                standard_error: array<float64>, \n",
      "                t_stat: array<float64>, \n",
      "                p_value: array<float64>, \n",
      "                multiple_standard_error: float64, \n",
      "                multiple_r_squared: float64, \n",
      "                adjusted_r_squared: float64, \n",
      "                f_stat: float64, \n",
      "                multiple_p_value: float64, \n",
      "                n: int32\n",
      "            }, \n",
      "            n_deletion: struct {\n",
      "                beta: array<float64>, \n",
      "                standard_error: array<float64>, \n",
      "                t_stat: array<float64>, \n",
      "                p_value: array<float64>, \n",
      "                multiple_standard_error: float64, \n",
      "                multiple_r_squared: float64, \n",
      "                adjusted_r_squared: float64, \n",
      "                f_stat: float64, \n",
      "                multiple_p_value: float64, \n",
      "                n: int32\n",
      "            }, \n",
      "            r_het_hom_var: struct {\n",
      "                beta: array<float64>, \n",
      "                standard_error: array<float64>, \n",
      "                t_stat: array<float64>, \n",
      "                p_value: array<float64>, \n",
      "                multiple_standard_error: float64, \n",
      "                multiple_r_squared: float64, \n",
      "                adjusted_r_squared: float64, \n",
      "                f_stat: float64, \n",
      "                multiple_p_value: float64, \n",
      "                n: int32\n",
      "            }, \n",
      "            n_transition: struct {\n",
      "                beta: array<float64>, \n",
      "                standard_error: array<float64>, \n",
      "                t_stat: array<float64>, \n",
      "                p_value: array<float64>, \n",
      "                multiple_standard_error: float64, \n",
      "                multiple_r_squared: float64, \n",
      "                adjusted_r_squared: float64, \n",
      "                f_stat: float64, \n",
      "                multiple_p_value: float64, \n",
      "                n: int32\n",
      "            }, \n",
      "            n_transversion: struct {\n",
      "                beta: array<float64>, \n",
      "                standard_error: array<float64>, \n",
      "                t_stat: array<float64>, \n",
      "                p_value: array<float64>, \n",
      "                multiple_standard_error: float64, \n",
      "                multiple_r_squared: float64, \n",
      "                adjusted_r_squared: float64, \n",
      "                f_stat: float64, \n",
      "                multiple_p_value: float64, \n",
      "                n: int32\n",
      "            }\n",
      "        }, \n",
      "        qc_metrics_stats: struct {\n",
      "            n_snp_residual: struct {\n",
      "                median: float64, \n",
      "                mad: float64, \n",
      "                lower: float64, \n",
      "                upper: float64\n",
      "            }, \n",
      "            n_singleton_residual: struct {\n",
      "                median: float64, \n",
      "                mad: float64, \n",
      "                lower: float64, \n",
      "                upper: float64\n",
      "            }, \n",
      "            r_ti_tv_residual: struct {\n",
      "                median: float64, \n",
      "                mad: float64, \n",
      "                lower: float64, \n",
      "                upper: float64\n",
      "            }, \n",
      "            r_insertion_deletion_residual: struct {\n",
      "                median: float64, \n",
      "                mad: float64, \n",
      "                lower: float64, \n",
      "                upper: float64\n",
      "            }, \n",
      "            n_insertion_residual: struct {\n",
      "                median: float64, \n",
      "                mad: float64, \n",
      "                lower: float64, \n",
      "                upper: float64\n",
      "            }, \n",
      "            n_deletion_residual: struct {\n",
      "                median: float64, \n",
      "                mad: float64, \n",
      "                lower: float64, \n",
      "                upper: float64\n",
      "            }, \n",
      "            r_het_hom_var_residual: struct {\n",
      "                median: float64, \n",
      "                mad: float64, \n",
      "                lower: float64, \n",
      "                upper: float64\n",
      "            }, \n",
      "            n_transition_residual: struct {\n",
      "                median: float64, \n",
      "                mad: float64, \n",
      "                lower: float64, \n",
      "                upper: float64\n",
      "            }, \n",
      "            n_transversion_residual: struct {\n",
      "                median: float64, \n",
      "                mad: float64, \n",
      "                lower: float64, \n",
      "                upper: float64\n",
      "            }\n",
      "        }, \n",
      "        n_pcs: int32, \n",
      "        used_regressed_metrics: bool\n",
      "    } \n",
      "    'gnomad_age_distribution': struct {\n",
      "        bin_edges: array<float64>, \n",
      "        bin_freq: array<int32>, \n",
      "        n_smaller: int32, \n",
      "        n_larger: int32\n",
      "    } \n",
      "    'variant_annotation_descriptions': struct {\n",
      "        locus: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        alleles: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        rsid: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        a_index: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        was_split: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        hgdp_tgp_freq: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                AC: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AF: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AN: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                homozygote_count: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        gnomad_freq: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                AC: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AF: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AN: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                homozygote_count: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        gnomad_popmax: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                AC: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AF: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AN: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                homozygote_count: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                pop: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                faf95: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        gnomad_faf: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                faf95: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                faf99: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        gnomad_qual_hists: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                gq_hist_all: struct {\n",
      "                    Description: str, \n",
      "                    sub_annotations: struct {\n",
      "                        bin_edges: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        bin_freq: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_smaller: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_larger: struct {\n",
      "                            Description: str\n",
      "                        }\n",
      "                    }\n",
      "                }, \n",
      "                dp_hist_all: struct {\n",
      "                    Description: str, \n",
      "                    sub_annotations: struct {\n",
      "                        bin_edges: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        bin_freq: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_smaller: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_larger: struct {\n",
      "                            Description: str\n",
      "                        }\n",
      "                    }\n",
      "                }, \n",
      "                gq_hist_alt: struct {\n",
      "                    Description: str, \n",
      "                    sub_annotations: struct {\n",
      "                        bin_edges: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        bin_freq: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_smaller: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_larger: struct {\n",
      "                            Description: str\n",
      "                        }\n",
      "                    }\n",
      "                }, \n",
      "                dp_hist_alt: struct {\n",
      "                    Description: str, \n",
      "                    sub_annotations: struct {\n",
      "                        bin_edges: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        bin_freq: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_smaller: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_larger: struct {\n",
      "                            Description: str\n",
      "                        }\n",
      "                    }\n",
      "                }, \n",
      "                ab_hist_alt: struct {\n",
      "                    Description: str, \n",
      "                    sub_annotations: struct {\n",
      "                        bin_edges: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        bin_freq: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_smaller: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_larger: struct {\n",
      "                            Description: str\n",
      "                        }\n",
      "                    }\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        gnomad_raw_qual_hists: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                gq_hist_all: struct {\n",
      "                    Description: str, \n",
      "                    sub_annotations: struct {\n",
      "                        bin_edges: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        bin_freq: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_smaller: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_larger: struct {\n",
      "                            Description: str\n",
      "                        }\n",
      "                    }\n",
      "                }, \n",
      "                dp_hist_all: struct {\n",
      "                    Description: str, \n",
      "                    sub_annotations: struct {\n",
      "                        bin_edges: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        bin_freq: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_smaller: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_larger: struct {\n",
      "                            Description: str\n",
      "                        }\n",
      "                    }\n",
      "                }, \n",
      "                gq_hist_alt: struct {\n",
      "                    Description: str, \n",
      "                    sub_annotations: struct {\n",
      "                        bin_edges: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        bin_freq: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_smaller: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_larger: struct {\n",
      "                            Description: str\n",
      "                        }\n",
      "                    }\n",
      "                }, \n",
      "                dp_hist_alt: struct {\n",
      "                    Description: str, \n",
      "                    sub_annotations: struct {\n",
      "                        bin_edges: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        bin_freq: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_smaller: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_larger: struct {\n",
      "                            Description: str\n",
      "                        }\n",
      "                    }\n",
      "                }, \n",
      "                ab_hist_alt: struct {\n",
      "                    Description: str, \n",
      "                    sub_annotations: struct {\n",
      "                        bin_edges: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        bin_freq: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_smaller: struct {\n",
      "                            Description: str\n",
      "                        }, \n",
      "                        n_larger: struct {\n",
      "                            Description: str\n",
      "                        }\n",
      "                    }\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        gnomad_age_hist_het: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                bin_edges: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                bin_freq: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_smaller: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_larger: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        gnomad_age_hist_hom: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                bin_edges: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                bin_freq: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_smaller: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_larger: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        filters: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        info: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                QUALapprox: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                SB: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                MQ: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                MQRankSum: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                VarDP: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AS_ReadPosRankSum: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AS_pab_max: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AS_QD: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AS_MQ: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                QD: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AS_MQRankSum: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                FS: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AS_FS: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                ReadPosRankSum: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AS_QUALapprox: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AS_SB_TABLE: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AS_VarDP: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AS_SOR: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                SOR: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                transmitted_singleton: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                omni: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                mills: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                monoallelic: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                InbreedingCoeff: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        vep: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        vqsr: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                AS_VQSLOD: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                AS_culprit: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                NEGATIVE_TRAIN_SITE: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                POSITIVE_TRAIN_SITE: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        region_flag: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                lcr: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                segdup: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        allele_info: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                variant_type: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                allele_type: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                n_alt_alleles: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        was_mixed: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        cadd: struct {\n",
      "            sub_annotations: struct {\n",
      "                raw_score: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                phred: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                has_duplicate: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        revel: struct {\n",
      "            Description: str, \n",
      "            sub_annotations: struct {\n",
      "                revel_score: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                has_duplicate: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        splice_ai: struct {\n",
      "            sub_annotations: struct {\n",
      "                splice_ai: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                splice_consequence: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                has_duplicate: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        primate_ai: struct {\n",
      "            sub_annotations: struct {\n",
      "                primate_ai_score: struct {\n",
      "                    Description: str\n",
      "                }, \n",
      "                has_duplicate: struct {\n",
      "                    Description: str\n",
      "                }\n",
      "            }\n",
      "        }, \n",
      "        AS_lowqual: struct {\n",
      "            Description: str\n",
      "        }, \n",
      "        telomere_or_centromere: struct {\n",
      "            Description: str\n",
      "        }\n",
      "    } \n",
      "    'hgdp_tgp_freq_meta': array<dict<str, str>> \n",
      "    'hgdp_tgp_freq_index_dict': dict<str, int32> \n",
      "    'gnomad_freq_meta': array<dict<str, str>> \n",
      "    'gnomad_freq_index_dict': dict<str, int32> \n",
      "    'gnomad_faf_index_dict': dict<str, int32> \n",
      "    'gnomad_faf_meta': array<dict<str, str>> \n",
      "    'vep_version': str \n",
      "    'vep_csq_header': str \n",
      "    'dbsnp_version': str \n",
      "    'variant_filtering_model': struct {\n",
      "        model_name: str, \n",
      "        score_name: str, \n",
      "        snv_cutoff: struct {\n",
      "            bin: float64, \n",
      "            min_score: float64\n",
      "        }, \n",
      "        indel_cutoff: struct {\n",
      "            bin: float64, \n",
      "            min_score: float64\n",
      "        }, \n",
      "        snv_training_variables: array<str>, \n",
      "        indel_training_variables: array<str>\n",
      "    } \n",
      "    'variant_inbreeding_coeff_cutoff': float64 \n",
      "----------------------------------------\n",
      "Row fields:\n",
      "    'genetic_region': str \n",
      "    'population': str \n",
      "    'n_snp_stats.mean': float64 \n",
      "    'n_snp_stats.stdev': float64 \n",
      "    'n_snp_stats.min': float64 \n",
      "    'n_snp_stats.max': float64 \n",
      "    'n_snp_stats.n': int64 \n",
      "    'n_snp_stats.sum': float64 \n",
      "    'cov_stats.mean': float64 \n",
      "    'cov_stats.stdev': float64 \n",
      "    'cov_stats.min': float64 \n",
      "    'cov_stats.max': float64 \n",
      "    'cov_stats.n': int64 \n",
      "    'cov_stats.sum': float64 \n",
      "    'n_unrelated': int64 \n",
      "----------------------------------------\n",
      "Key: ['genetic_region', 'population']\n",
      "----------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# Check the description of the table after flattening to make sure it is what we'd expect\n",
    "table.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "f0a6fd0c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-11-18 21:10:47 Hail: INFO: Ordering unsorted dataset with network shuffle\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "78"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# One last validity check before writing out the dataset is to make sure we still have the number of rows we'd expect\n",
    "# In this case, since the data is grouped by genetic region, population, the number of rows should be equal to the number of populations (78)\n",
    "table.count()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04309c36",
   "metadata": {},
   "source": [
    "# 7. Export Final Table\n",
    "<br>\n",
    "<details><summary>For more information on Hail methods and expressions click <u><span style=\"color:blue\">here</span></u>.</summary> \n",
    "<br>\n",
    "<li><a href=\"https://hail.is/docs/0.2/hail.Table.html#hail.Table.export\"> More on  <i> export() </i></a></li>\n",
    "\n",
    "</details>\n",
    "\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3496f63d",
   "metadata": {},
   "source": [
    "- Write out the final table in tsv format \n",
    "\n",
    "```python3\n",
    "table.export(final_table_path, header=True)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a33a10e9",
   "metadata": {},
   "source": [
    "[Back to Index](#Index)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  },
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
